File: src/utils/log_user_action.py
# 테이블 상태 스냅샷 기록 (토글 켤 때 BEFORE, 끌 때 AFTER)
def log_table_edit_snapshot(table_name: str, state: dict, when: str):
    """
    table_name: e.g. 'PERCEPTION-SECTION', 'PERCEPTION-COMPONENT', etc.
    state: 전체 테이블 데이터 (dict or list)
    when: 'BEFORE' (토글 켤 때), 'AFTER' (토글 끌 때)
    """
    log_user_action('edit_table_snapshot', table_name, {
        "when": when,
        "state": state
    })
# 단계별 결과 기록
def log_step_result(step: str, task: str, image_path: str, result: str):
    log_user_action('step_result', step, {"task": task, "imagePath": image_path, "result": result})

import os
import json
from fastapi import Request
from fastapi.responses import JSONResponse
from src.constants import USER_ID

def log_user_action(action_type: str, step_info: str, details: dict = None):
    log_entry = {
        "timestamp": __import__('datetime').datetime.utcnow().isoformat() + 'Z',
        "action_type": action_type,
        "step": step_info,
        "detail": details or {}
    }
    log_dir = os.path.join(os.path.dirname(__file__), '../../logs')
    os.makedirs(log_dir, exist_ok=True)
    log_file = os.path.join(log_dir, f'user_{USER_ID}_actions.log')
    with open(log_file, 'a') as f:
        f.write(json.dumps(log_entry, ensure_ascii=False) + '\n')


# 테이블 수정 시작 (토글 켜기)
def log_table_edit_start(table_name: str):
    log_user_action('edit_table_start', table_name, {})

# 테이블 필드 수정 기록
def log_table_field_edit(table_name: str, row_id: str, field: str, before: str, after: str):
    log_user_action('edit_table_field', table_name, {
        "rowId": row_id,
        "field": field,
        "before": before,
        "after": after
    })

# 테이블 수정 종료 (토글 끄기)
def log_table_edit_end(table_name: str):
    log_user_action('edit_table_end', table_name, {})

# 가이드라인 수정 프롬프트 기록
def log_guideline_edit_prompt(prompt: str, before_guideline: str, after_guideline: str):
    log_user_action('guideline_edit_prompt', prompt, {"beforeGuideline": before_guideline, "afterGuideline": after_guideline})

# 가이드라인 실제 수정 내역 기록
def log_guideline_updated(before_guideline: str, after_guideline: str):
    log_user_action('guideline_updated', 'Guideline updated', {"beforeGuideline": before_guideline, "afterGuideline": after_guideline})

# FastAPI endpoint example
from fastapi import APIRouter
router = APIRouter()

@router.post('/log-action')
async def log_action(request: Request):
    data = await request.json()
    user_id = data.get('userId', 'anonymous')
    action_type = data.get('action_type')
    step_info = data.get('step_info')
    details = data.get('details', {})
    log_user_action(user_id, action_type, step_info, details)
    return JSONResponse({"success": True})

File: src/utils/logUserAction.js
// 테이블 상태 스냅샷 기록 (토글 켤 때 BEFORE, 끌 때 AFTER)
export async function logTableEditSnapshot({ tableName, state, when }) {
    await logUserAction({
        action_type: 'edit_table_snapshot',
        content: tableName,
        details: { when, state }
    });
}

// API base URL ('' for same-origin in production). In dev, set NEXT_PUBLIC_API_BASE=http://localhost:8000
// If not set and running on localhost, default to FastAPI port 8000
const API_BASE = (typeof window !== 'undefined' && !process.env.NEXT_PUBLIC_API_BASE && window.location.hostname === 'localhost')
    ? 'http://34.64.194.66:8000'
    : (process.env.NEXT_PUBLIC_API_BASE || '');

// API endpoint for logging
const LOG_API_URL = `${API_BASE}/api/log-user-action/`;

export async function logUserAction({ action_type, content, details = {} }) {
    const logEntry = {
        timestamp: new Date().toISOString(),
        // userId,
        action_type,
        content,
        details
    };
    try {
        let response = await fetch(LOG_API_URL, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(logEntry)
        });
        if (!response.ok) {
            // Try fallback without trailing slash once
            const bodyText = await response.text().catch(() => '');
            console.error('Failed to send user action log:', response.status, bodyText);
            if (response.status === 404 || response.status === 405) {
                try {
                    response = await fetch(`${API_BASE}/api/log-user-action`, {
                        method: 'POST',
                        headers: { 'Content-Type': 'application/json' },
                        body: JSON.stringify(logEntry)
                    });
                    if (!response.ok) {
                        const retryText = await response.text().catch(() => '');
                        console.error('Retry failed (no-slash):', response.status, retryText);
                    }
                } catch (e) {
                    console.error('Retry error (no-slash):', e);
                }
            }
        }
    } catch (err) {
        console.error('Failed to send user action log:', err);
    }
}

export async function logUserPrompt({ userId = 'anonymous', prompt, relatedTask }) {
    await logUserAction({
        userId,
        action_type: 'user_prompt',
        content: prompt,
        details: { relatedTask }
    });
}

export async function logStepResult({ userId = 'anonymous', step, task, imagePath, result, duration }) {
    await logUserAction({
        userId,
        action_type: 'step_result',
        content: step,
        details: { task, imagePath, result, duration }
    });
}

export async function logGuidelineEditPrompt({ userId = 'anonymous', step, prompt, beforeGuideline, afterGuideline }) {
    await logUserAction({
        userId,
        action_type: 'guideline_edit_prompt',
        content: step,
        details: { prompt, beforeGuideline, afterGuideline }
    });
}

export async function logGuidelineUpdated({ userId = 'anonymous', beforeGuideline, afterGuideline }) {
    await logUserAction({
        userId,
        action_type: 'guideline_updated',
        content: 'Guideline updated',
        details: { beforeGuideline, afterGuideline }
    });
}
